streams/src/main/java/org/apache/kafka/streams/kstream/type/Resolver.java

```diff
@@ -0,0 +1,133 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.streams.kstream.type;
+
+import org.apache.kafka.common.KafkaException;
+import org.apache.kafka.streams.kstream.Windowed;
+
+import java.lang.reflect.Method;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.lang.reflect.TypeVariable;
+
+public class Resolver {
+
+    /**
+     * Construct an instance of TypeWithTypeArgs
+     */
+    public static TypeWithTypeArgs getTypeWithTypeArgs(Class<?> type, Type[] typeArgs) throws TypeException {
+        verify(type, typeArgs);
+
+        return new TypeWithTypeArgs(type, typeArgs);
+    }
+
+    /**
+     * Resolves the type from an method return type. This returns null when there is not sufficient information to
+     * determine fully resolved type.
+     */
+    public static Type resolveFromMethodReturnType(Object obj, String methodName, Class... argTypes) {
```

Comment:
not right now. I initially use this to infer types for map functions. But there is an issue, so I removed the code using this. I will remove it once I get a better solution.